### PR TITLE
Adaptive ephemerality

### PR DESCRIPTION
**Pull request by davidmarkclements**
in order to support this through the stack every layer needed an update:

* https://github.com/mafintosh/dht-rpc/pull/9
* https://github.com/hyperswarm/dht/pull/27/
* https://github.com/hyperswarm/discovery/pull/15
* https://github.com/hyperswarm/network/pull/18
* https://github.com/hyperswarm/hyperswarm/pull/51

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,13 +5,14 @@
   "main": "index.js",
   "dependencies": {
     "@hyperswarm/hypersign": "^2.0.0",
-    "dht-rpc": "^4.2.0",
+    "dht-rpc": "git+https://github.com/davidmarkclements/dht-rpc.git#turn-non-ephemeral",
```
Comment (davidmarkclements) [inline package.json]: TODO: update to next dht-rpc release when https://github.com/mafintosh/dht-rpc/pull/9 lands

File: index.js [hunk 2]
```diff
@@ -47,6 +49,26 @@ class HyperDHT extends DHT {
...
+          this.joinDht((err) => {
+            if (err) {
+              err.message = `Unable to dynamically become non-ephemeral: ${err.message}`
+              this.emit('warning', err)
+              return
+            }
+            this.emit('dynamically-non-ephemeral')
+          })
+        }, ADAPT_EPHEMERALITY_AFTER)
+        timeout.unref()
```
Comment (mafintosh) [inline index.js]: we should clear the timeout on destroy

Comment (mafintosh) [reply index.js]: oh and we should check if we're holepunchable also before going persistent

File: index.js [hunk 3]
```diff
@@ -47,6 +49,26 @@ class HyperDHT extends DHT {
...
+        const timeout = setTimeout(() => {
+          this.joinDht((err) => {
+            if (err) {
+              err.message = `Unable to dynamically become non-ephemeral: ${err.message}`
+              this.emit('warning', err)
+              return
+            }
+            this.emit('dynamically-non-ephemeral')
+          })
+        }, ADAPT_EPHEMERALITY_AFTER)
```
Comment (mafintosh) [inline index.js]: would prob be good to add some randomness here, to avoid cascading joins

Comment (mafintosh) [reply index.js]: ie EPH_AFTER + Math.random() * EPH_AFTER / 2 or something

Comment (davidmarkclements) [reply index.js]: done :+1:

File: index.js (hunk 2 above)
Comment (davidmarkclements) [reply index.js]: done!

File: index.js [hunk 4]
```diff
@@ -44,9 +47,64 @@ class HyperDHT extends DHT {
...
+
+    if (opts.adaptive) {
+      if (this.ephemeral !== true) {
+        this.destroy()
+        throw Error('adaptive mode can only applied when ephemeral: true')
+      }
+      this.once('ready', () => {
+        this._adaptiveTimeout = setTimeout(() => {
+          this.holepunchable((err, able) => {
+            if (!err && able) {
```
Comment (mafintosh) [inline index.js]: if (err || !able) tryAgainInDoubleEphAfterIsh

File: index.js [hunk 5]
```diff
@@ -44,9 +47,64 @@ class HyperDHT extends DHT {
...
+        for (var i = 1; i < res.length; i++) {
+          const pong = res[i].pong
+          if (pong.host !== first.host || pong.port !== first.port) {
+            return cb(null, false)
+          }
+        }
+
+        cb(null, true)
+      })
+    }
```
Comment (mafintosh) [inline index.js]: excellent! important next thing would be to make this work with non bootstraps

File: index.js (hunk 4 above)
Comment (mafintosh) [reply index.js]: if not closed

File: index.js (hunk 5 above)
Comment (davidmarkclements) [reply index.js]: 💯

File: package.json [hunk 6]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "dependencies": {
     "@hyperswarm/hypersign": "^2.0.0",
-    "dht-rpc": "^4.2.0",
+    "dht-rpc": "git+https://github.com/davidmarkclements/dht-rpc.git#persistent",
```
Comment (davidmarkclements) [inline package.json]: TODO: update when https://github.com/mafintosh/dht-rpc/pull/19 lands

File: index.js [hunk 7]
```diff
@@ -44,9 +47,33 @@ class HyperDHT extends DHT {
...
+
+    this._adaptiveTimeout = null
+
+    if (opts.adaptive) {
+      if (this.ephemeral !== true) {
+        this.destroy()
+        throw Error('adaptive mode can only applied when ephemeral: true')
+      }
+      this.once('ready', () => {
+        this._adaptiveTimeout = setTimeout(() => {
```
Comment (mafintosh) [inline index.js]: we should move this to use guard-timeout if trivial now :) otherwise we can do so later

Comment (davidmarkclements) [reply index.js]: added

File: package.json (hunk 6 above)
Comment (davidmarkclements) [reply package.json]:
```suggestion
    "dht-rpc": ^4.7.0",
```

**Comment by mafintosh** [comment]
3.3.0

**Pull request merged**